### PR TITLE
[OSDOCS-3716]: Add etcd hardware recommendations

### PR DESCRIPTION
**Pull request by tmalove**
Version(s): 4.10+

Issue: [[OSDOCS-3716](https://issues.redhat.com//browse/OSDOCS-3716) (subtask of [OCDOCS-3589](https://issues.redhat.com/browse/OSDOCS-3589))

(UPDATED - 7/12/2022) Link to docs preview: http://file.rdu.redhat.com/tlove/etcd-3716-hardware-reco/scalability_and_performance/recommended-host-practices.html#recommended-etcd-practices_recommended-host-practices

### CODE REVIEW

**Comment by tjungblu** [comment]
Thanks @tmalove I'll check this first thing tomorrow morning :) 
I think the KCS is this one: https://access.redhat.com/solutions/4885641

File: modules/recommended-etcd-practices.adoc [hunk 1]
```diff
@@ -18,6 +18,20 @@ For more information about defragmenting etcd, see the "Defragmenting etcd data"
...
 Because etcd writes data to disk and persists proposals on disk, its performance depends on disk performance. Slow disks and disk activity from other processes can cause long fsync latencies. Those latencies can cause etcd to miss heartbeats, not commit new proposals to the disk on time, and ultimately experience request timeouts and temporary leader loss. Run etcd on machines that are backed by SSD or NVMe disks with low latency and high throughput. Consider single-level cell (SLC) solid-state drives (SSDs), which provide 1 bit per memory cell, are durable and reliable, and are ideal for write-intensive workloads.
 
+The following hard disk features support optimal etcd performance:
+
+* Low latency to support fast read operation.
+* High-bandwidth writes for faster compactions and defragmentation.
+* High-bandwidth reads for faster recovery from failures.
+* Solid state drives as a minimum selection, however NVME drives are preferred.
+* Server-grade hardware from various manufacturers for increased reliability.
+* RAID 0 or RAID 1 configurations for reliable data.
```
Comment (tjungblu) [inline modules/recommended-etcd-practices.adoc:28]: nit: RAID 0 is for performance, RAID 1 for reliability. I would stress that RAID 1 is not necessary in the context of etcd, as your data is already replicated over three nodes.

File: modules/recommended-etcd-practices.adoc [hunk 2]
```diff
@@ -18,6 +18,20 @@ For more information about defragmenting etcd, see the "Defragmenting etcd data"
...
+* Dedicated etcd drives for heavy input-output workloads, such as log files.
```
Comment (tjungblu) [inline modules/recommended-etcd-practices.adoc:29]: the meaning I wanted to convey here is that customers should **not** co-locate the log files on the same drives that etcd uses. Can you rework that sentence to make it clear please?

File: modules/recommended-etcd-practices.adoc [hunk 3]
```diff
@@ -18,6 +18,20 @@ For more information about defragmenting etcd, see the "Defragmenting etcd data"
...
+
+Avoid network or spinning drives. Always benchmark using utilities such as `fio` or _etcd troubleshooting tools on {product-title}_.
```
Comment (tjungblu) [inline modules/recommended-etcd-practices.adoc:31]:
```suggestion
Avoid NAS or SAN setups and spinning drives. Always benchmark using utilities such as `fio` or _etcd troubleshooting tools on {product-title}_. And continuously monitor the performance of your cluster as it grows.
```

**Comment by tjungblu** [comment]
@tmalove done :) looks great already, just added couple of nitty comments on some statements

**Comment by tmalove** [comment]
Thanks @tjungblu! I just added the updated preview also.

**Comment by tmalove** [comment]
@geliu2016  Will you verify these updates for etcd hardware recommendations? Thanks!

**Comment by geliu2016** [comment]
Hi @tjungblu 
is it possible provide some etcd prerequisite about master node's storage/cpu/memory in different platform(aws, gcp, vsphere,...)? as I know, some customer or CEE team member have interest about this paremeters for etcd. thanks

File: scalability_and_performance/recommended-host-practices.adoc [hunk 4]
```diff
@@ -29,6 +29,11 @@ include::modules/increasing-aws-flavor-size.adoc[leveloffset=+2]
 
 include::modules/recommended-etcd-practices.adoc[leveloffset=+1]
 
+[role="_additional-resources"]
+.Additional resources
```
Comment (jldohmann) [inline scalability_and_performance/recommended-host-practices.adoc:33]: Needs an ID: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies

Suggested edit:
```suggestion
[id="additional-resources_recommended-etcd-practices"]
.Additional resources
```

File: modules/recommended-etcd-practices.adoc [hunk 5]
```diff
@@ -18,6 +18,20 @@ For more information about defragmenting etcd, see the "Defragmenting etcd data"
...
 
+The following hard disk features provide optimal etcd performance:
+
+* Low latency to support fast read operation.
+* High-bandwidth writes for faster compactions and defragmentation.
+* High-bandwidth reads for faster recovery from failures.
+* Solid state drives as a minimum selection, however NVME drives are preferred.
+* Server-grade hardware from various manufacturers for increased reliability.
+* RAID 0 technology for increased performance.
+* Dedicated etcd drives. Do not locate log files or other heavy workloads on etcd drives.
```
Comment (jldohmann) [inline modules/recommended-etcd-practices.adoc:29]: The word locate sounds kind of strange in this way. Maybe "place" would read better here, what do you think? 
```suggestion
* Dedicated etcd drives. Do not place log files or other heavy workloads on etcd drives.
```

**Review by jldohmann — COMMENTED** [review]
It's looking good! Some minor things plus a question 😄

**Review by jldohmann — APPROVED** [review]

File: scalability_and_performance/recommended-host-practices.adoc (hunk 4 above)
Comment (tmalove) [reply scalability_and_performance/recommended-host-practices.adoc:34]: > Needs an ID: [https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc?rgh-link-date=2022-07-07T17%3A09%3A55Z#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies)
> 
> Suggested edit:

Hi @jldohmann Thanks for your comments! For this one, the 'role=' statements is needed, yes? That is what I see in the manual.

Comment (jldohmann) [reply scalability_and_performance/recommended-host-practices.adoc:34]: @tmalove yep! Both the role and ID are needed

File: scalability_and_performance/recommended-host-practices.adoc [hunk 6]
```diff
@@ -29,6 +29,12 @@ include::modules/increasing-aws-flavor-size.adoc[leveloffset=+2]
 
 include::modules/recommended-etcd-practices.adoc[leveloffset=+1]
 
+[role="_additional-resources"]
+[id="additional-resources_recommended-etcd-practices"]
+.Additional resources
+* link:https://github.com/peterducai/openshift-etcd-suite[etcd troubleshooting tools on {product-title}]
```
Comment (michaelryanpeter) [inline scalability_and_performance/recommended-host-practices.adoc:35]: Linking to documentation and resources hosted on GitHub pages outside of the `openshift` organization is not allowed. Occasionally, exceptions are made but those need to go our change management process. In most cases, the exceptions that are approved tend to be hosted in an official upstream repo. 

I suggest cutting this link from the current PR and revisiting it in a follow up ticket so you don't hold up the rest of the content.

**Review by michaelryanpeter — COMMENTED** [review]
Please let me know if you have questions.

File: scalability_and_performance/recommended-host-practices.adoc (hunk 6 above)
Comment (tmalove) [reply scalability_and_performance/recommended-host-practices.adoc]: Done.

**Comment by tmalove** [comment]
@michaelryanpeter I made the updates, however the preview still shows the Github link reference in the module and in the Additional resource. The only other thing that I did was to remove the ID from the trialing include...do you see anything glaring? I will keep looking also.

**Pull request merged**